File: src/backend/app/services/rotation_safe_area.py
"""
Rotation safe-area geometry (T5640 — Framing horizon straighten) — Python mirror.

This is the backend copy of `src/frontend/src/utils/rotationSafeArea.js`.

2026-09-18 data-loss fix: this used to be a defense-only mirror -- the CLIENT
clamped and PERSISTED crop keyframes the moment the straighten angle changed,
so the export trusted the stored (already-clamped) crop and never re-clamped
at render time. That write-time clamp had no inverse and destroyed the user's
tracking data (see useCrop.setRotation's docstring on the frontend for the
mechanism). Crop keyframes now store the user's true framing verbatim; the
safe-area clamp moved HERE, to render time, which is now the ONLY place that
guarantees no black wedge enters the export. Callers: this module directly
from `app.ai_upscaler.frame_processor` (same process, can import `app`);
`app/modal_functions/video_processing.py` keeps its OWN inline copy
(`_clamp_crop_to_safe_area`) since its Modal image can't import `app`. Keep
all three (this file, the inline Modal copy, and rotationSafeArea.js) in sync.

Sign convention (design §2.1): theta = content-correction angle in degrees,
positive = rotate content counter-clockwise (math orientation, y-up). The render
is always crop(rotate_theta_aboutCenter_sameWH(frame)); theta=0 is identity.
"""

import math

# Hard rotation cap in degrees (mirrors MAX_ROT in the JS util).
MAX_ROT = 20

# Rotation dead-zone in degrees (T6170) — mirrors ROTATION_EPSILON in the JS util.
# Any |theta| below this is FP residue from trig / repeated 0.1-degree dial
# additions (0.1+0.1+0.1-0.1-0.1-0.1 == 2.7755575615628914e-17), NOT a real angle,
# and must be treated as theta=0. 1e-6 is 100,000x below the finest 0.1-degree
# user step and >=5 orders of magnitude above any accumulation residue. 2026-09-18:
# this module is now in the render path (frame_processor.py imports it directly),
# so this dead-zone is live production behavior, not just defense-in-depth for a
# characterization test.
ROTATION_EPSILON = 1e-6


def max_axis_aligned_in_rotated(w: float, h: float, theta_deg: float) -> tuple[float, float]:
    """Largest axis-aligned rectangle (any aspect), centered, inside a W*H frame
    rotated by theta degrees. Returns (width, height) in frame pixels."""
    a = abs(theta_deg) * math.pi / 180
    if a == 0:
        return (w, h)

    sin_a = abs(math.sin(a))
    cos_a = abs(math.cos(a))
    longer = max(w, h)
    shorter = min(w, h)
    width_is_longer = w >= h

    if shorter <= 2 * sin_a * cos_a * longer or abs(sin_a - cos_a) < 1e-10:
        half_short = 0.5 * shorter
        if width_is_longer:
            wr = half_short / sin_a
            hr = half_short / cos_a
        else:
            wr = half_short / cos_a
            hr = half_short / sin_a
    else:
        cos_2a = cos_a * cos_a - sin_a * sin_a
        wr = (w * cos_a - h * sin_a) / cos_2a
        hr = (h * cos_a - w * sin_a) / cos_2a
    return (wr, hr)


def safe_area_for_aspect(w: float, h: float, theta_deg: float, r: float) -> dict:
    """Largest centered box of target aspect r that fits inside the inscribed
    rectangle for (w, h, theta). Returns {x0, y0, w_safe, h_safe}."""
    wr, hr = max_axis_aligned_in_rotated(w, h, theta_deg)

    if wr / hr >= r:
        w_safe = hr * r
        h_safe = hr
    else:
        w_safe = wr
        h_safe = wr / r

    x0 = (w - w_safe) / 2
    y0 = (h - h_safe) / 2
    return {"x0": x0, "y0": y0, "w_safe": w_safe, "h_safe": h_safe}


def clamp_crop_to_safe_area(crop: dict, w: float, h: float, theta_deg: float, r: float) -> dict:
    """Clamp a crop {x, y, width, height} to the inscribed safe area for
    (w, h, theta), preserving aspect r exactly. |theta| < ROTATION_EPSILON is an
    identity fast path (T6170: a denormal must not engage the clamp). `not theta_deg`
    is kept as the leading term so None/0 passthrough is unchanged. Returns
    {x, y, width, height}."""
    if not theta_deg or abs(theta_deg) < ROTATION_EPSILON:
        return {"x": crop["x"], "y": crop["y"], "width": crop["width"], "height": crop["height"]}

    s = safe_area_for_aspect(w, h, theta_deg, r)

    cw = min(crop["width"], s["w_safe"])
    ch = min(crop["height"], s["h_safe"])
    if cw / ch > r:
        cw = ch * r
    else:
        ch = cw / r

    max_x = s["x0"] + s["w_safe"] - cw
    max_y = s["y0"] + s["h_safe"] - ch
    x = min(max(crop["x"], s["x0"]), max_x)
    y = min(max(crop["y"], s["y0"]), max_y)

    return {"x": x, "y": y, "width": cw, "height": ch}

File: src/frontend/src/utils/rotationSafeArea.js
/**
 * Rotation safe-area geometry (T5640 — Framing horizon straighten).
 *
 * SINGLE SOURCE OF TRUTH for the inscribed "no-black-corners" crop clamp and
 * the rotated-frame quad used by the out-of-bounds dim mask. Mirrored in Python
 * at `src/backend/app/services/rotation_safe_area.py` (kept in sync; the backend
 * copy exists only for the characterization test / defense — the export TRUSTS
 * the stored clamped crop and never re-clamps at render time).
 *
 * Coordinate / sign convention (design §2.1, pinned):
 *   theta = content-correction angle in DEGREES, positive = rotate content
 *   counter-clockwise (standard math orientation, y-up). Stored in
 *   working_clips.rotation. The RENDER is always:
 *       render(frame) = crop_{x,y,w,h}( rotate_theta_aboutCenter_sameWH(frame) )
 *   so crop coords live in the ROTATED frame space and theta=0 is byte-identical
 *   to today. Rotation is about the frame center with output size == source W*H,
 *   which preserves the coordinate box.
 *
 *   - CSS preview:  transform: rotate(-theta deg) about the display-rect center.
 *   - cv2 export:   getRotationMatrix2D(center, angle=theta)  (cv2 CCW positive).
 *   - ffmpeg:       rotate=a=-theta*PI/180 : ow=iw : oh=ih : c=black.
 */

// Hard rotation cap in degrees (design decision #3: user chose 20 over 15).
// The straighten tool and the fine dial both clamp to +/- this.
export const MAX_ROT = 20;

/**
 * Largest axis-aligned rectangle (unconstrained aspect) that fits, centered,
 * inside a W*H frame rotated by theta degrees. Classic rotatedRectWithMaxArea.
 * Returns { width, height } in frame pixels.
 */
export function maxAxisAlignedInRotated(W, H, thetaDeg) {
  const a = Math.abs(thetaDeg) * Math.PI / 180;
  if (a === 0) return { width: W, height: H };

  const sinA = Math.abs(Math.sin(a));
  const cosA = Math.abs(Math.cos(a));
  const longer = Math.max(W, H);
  const shorter = Math.min(W, H);
  const widthIsLonger = W >= H;

  let wr;
  let hr;
  if (shorter <= 2 * sinA * cosA * longer || Math.abs(sinA - cosA) < 1e-10) {
    // Half-constrained case: the rectangle touches the midpoint of the shorter side.
    const halfShort = 0.5 * shorter;
    if (widthIsLonger) {
      wr = halfShort / sinA;
      hr = halfShort / cosA;
    } else {
      wr = halfShort / cosA;
      hr = halfShort / sinA;
    }
  } else {
    const cos2a = cosA * cosA - sinA * sinA;
    wr = (W * cosA - H * sinA) / cos2a;
    hr = (H * cosA - W * sinA) / cos2a;
  }
  return { width: wr, height: hr };
}

/**
 * The largest centered box of target aspect r (= ratioW/ratioH) that fits inside
 * the inscribed rectangle for (W, H, theta). Returns the allowed crop region:
 * { x0, y0, wSafe, hSafe } in frame pixels.
 */
export function safeAreaForAspect(W, H, thetaDeg, r) {
  const { width: wr, height: hr } = maxAxisAlignedInRotated(W, H, thetaDeg);

  let wSafe;
  let hSafe;
  if (wr / hr >= r) {
    // Height-constrained.
    wSafe = hr * r;
    hSafe = hr;
  } else {
    // Width-constrained.
    wSafe = wr;
    hSafe = wr / r;
  }

  const x0 = (W - wSafe) / 2;
  const y0 = (H - hSafe) / 2;
  return { x0, y0, wSafe, hSafe };
}

/**
 * Clamp a crop box to the inscribed safe area for (W, H, theta), preserving the
 * target aspect r exactly. Shrinks (aspect-locked) then re-centers within the
 * allowed region. Deterministic — the guarantee that the export has no black
 * corners. Returns { x, y, width, height }.
 *
 * theta === 0 is an explicit identity fast path (no clamp) so the theta=0
 * behavior is byte-identical to today.
 */
export function clampCropToSafeArea(crop, W, H, thetaDeg, r) {
  if (!thetaDeg) return { x: crop.x, y: crop.y, width: crop.width, height: crop.height };

  const S = safeAreaForAspect(W, H, thetaDeg, r);

  let w = Math.min(crop.width, S.wSafe);
  let h = Math.min(crop.height, S.hSafe);
  // Keep aspect exactly r after the shrink.
  if (w / h > r) w = h * r;
  else h = w / r;

  const maxX = S.x0 + S.wSafe - w;
  const maxY = S.y0 + S.hSafe - h;
  const x = Math.min(Math.max(crop.x, S.x0), maxX);
  const y = Math.min(Math.max(crop.y, S.y0), maxY);

  return { x, y, width: w, height: h };
}

/**
 * The 4 corners of a W*H frame after the CSS content rotation, in FRAME-space
 * pixels (map each through videoToScreen to draw the dim-mask hole). Order:
 * top-left, top-right, bottom-right, bottom-left.
 *
 * The video element is CSS-rotated by rotate(-theta) about the display-rect
 * center; frame space and screen space share the y-down orientation and
 * videoToScreen is a uniform scale + offset, so rotating the corners about the
 * frame center by the same CSS angle here and then mapping is equivalent to the
 * on-screen rotated video. Uses the CSS rotate(a) matrix with a = -theta:
 *   x' = x cos a - y sin a ; y' = x sin a + y cos a   (about the center).
 */
export function rotatedFrameCorners(W, H, thetaDeg) {
  const a = -thetaDeg * Math.PI / 180;
  const cosA = Math.cos(a);
  const sinA = Math.sin(a);
  const cx = W / 2;
  const cy = H / 2;
  const rot = (px, py) => {
    const dx = px - cx;
    const dy = py - cy;
    return {
      x: cx + dx * cosA - dy * sinA,
      y: cy + dx * sinA + dy * cosA,
    };
  };
  return [rot(0, 0), rot(W, 0), rot(W, H), rot(0, H)];
}
